File: frontend/src/api/xp.ts
import { api } from './client';

export interface XpBadge {
  id: string;
  name: string;
  description: string;
  icon: string;
  earned: boolean;
  earned_at: string | null;
}

export interface XpSummary {
  total_xp: number;
  level: number;
  level_title: string;
  xp_in_level: number;
  xp_for_next_level: number;
  today_xp: number;
  today_max_xp: number;
  streak_days: number;
  weekly_xp: number;
  recent_badges: XpBadge[];
}

export interface XpHistoryItem {
  id: number;
  xp: number;
  reason: string;
  created_at: string;
}

export interface XpHistoryResponse {
  items: XpHistoryItem[];
  total: number;
}

export interface XpLedgerEntry {
  action_type: string;
  xp_awarded: number;
  multiplier: number;
  reason: string | null;
  created_at: string;
}

export interface XpLedgerResponse {
  entries: XpLedgerEntry[];
  total_count: number;
}

export interface BadgeResponse {
  badge_id: string;
  badge_name: string;
  badge_description: string;
  earned: boolean;
  awarded_at: string | null;
}

export interface XpStreakResponse {
  current_streak: number;
  longest_streak: number;
  streak_start_date: string | null;
}

export interface BrowniePointResponse {
  awarded: number;
  student_user_id: number;
  new_total_xp: number;
  remaining_weekly_cap: number;
  message: string;
}

export interface BrownieRemaining {
  remaining: number;
  weekly_cap: number;
}

export const xpApi = {
  getSummary: async () => {
    const response = await api.get<XpSummary>('/api/xp/summary');
    return response.data;
  },

  getHistory: async (limit = 50, offset = 0) => {
    const response = await api.get<XpHistoryResponse>(`/api/xp/history?limit=${limit}&offset=${offset}`);
    return response.data;
  },

  getLedger: async (limit = 50, offset = 0) => {
    const response = await api.get<XpLedgerResponse>(`/api/xp/history?limit=${limit}&offset=${offset}`);
    return response.data;
  },

  getBadges: async () => {
    const response = await api.get<XpBadge[]>('/api/xp/badges');
    return response.data;
  },

  getStreak: async () => {
    const response = await api.get<XpStreakResponse>('/api/xp/streak');
    return response.data;
  },

  getChildStreak: async (studentId: number) => {
    const response = await api.get<{ current_streak: number; longest_streak: number; freeze_tokens_remaining: number; streak_tier: string; multiplier: number; tier_label: string; last_streak_date: string | null }>(`/api/students/${studentId}/streak`);
    return response.data;
  },

  getChildSummary: async (studentId: number) => {
    const response = await api.get<XpSummary>(`/api/xp/children/${studentId}/summary`);
    return response.data;
  },

  awardBrowniePoints: async (studentUserId: number, points: number, reason?: string) => {
    const response = await api.post<BrowniePointResponse>('/api/xp/award', {
      student_user_id: studentUserId,
      points,
      reason: reason || undefined,
    });
    return response.data;
  },

  getBrownieRemaining: async (studentUserId: number) => {
    const response = await api.get<BrownieRemaining>(`/api/xp/award/remaining?student_user_id=${studentUserId}`);
    return response.data;
  },
};

File: app/schemas/xp.py
"""Pydantic schemas for the XP / Gamification system."""
from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class XpSummaryResponse(BaseModel):
    """Current user's XP summary."""
    user_id: int = 0
    total_xp: int = 0
    # Alias of total_xp for the frontend XpStreakBadge (#4019)
    xp_total: int = 0
    level: int = 1
    current_level: int = 1
    level_title: str = "Curious Learner"
    current_streak: int = 0
    longest_streak: int = 0
    freeze_tokens_remaining: int = 1
    xp_to_next_level: int = 200
    today_xp: int = 0
    today_cap: int = 0
    # Fields expected by the frontend XpSummary interface
    streak_days: int = 0
    xp_in_level: int = 0
    xp_for_next_level: int = 200
    today_max_xp: int = 0
    weekly_xp: int = 0
    recent_badges: list[dict[str, Any]] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class XpLedgerEntry(BaseModel):
    """Single XP ledger row."""
    action_type: str
    xp_awarded: int
    multiplier: float = 1.0
    reason: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class XpHistoryResponse(BaseModel):
    """Paginated XP history."""
    items: list[XpLedgerEntry] = Field(default_factory=list)
    total: int = 0
    limit: int = 50
    offset: int = 0

    @property
    def entries(self) -> list[XpLedgerEntry]:
        """Alias for items (backward compat with service-layer tests)."""
        return self.items

    @property
    def total_count(self) -> int:
        """Alias for total (backward compat with service-layer tests)."""
        return self.total


class BadgeResponse(BaseModel):
    """Badge info (earned or unearned)."""
    badge_id: str
    badge_name: str
    badge_description: str
    earned: bool = False
    awarded_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class StreakResponse(BaseModel):
    """Current streak info."""
    current_streak: int = 0
    longest_streak: int = 0
    freeze_tokens_remaining: int = 1
    multiplier: float = 1.0
    tier: str = "grey"
    streak_tier: Optional[str] = None
    tier_label: Optional[str] = None
    last_streak_date: Optional[str] = None
    last_activity_date: Optional[str] = None


class BrowniePointRequest(BaseModel):
    """Parent/Teacher awards brownie points to a student."""
    student_user_id: int
    points: int = Field(ge=1, le=50, description="Points to award (1-50)")
    reason: Optional[str] = Field(None, max_length=200)


class BrowniePointResponse(BaseModel):
    """Response after awarding brownie points."""
    awarded: int
    student_user_id: int
    new_total_xp: int
    remaining_weekly_cap: int = 0
    message: str
